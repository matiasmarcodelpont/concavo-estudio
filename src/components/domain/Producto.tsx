'use client'

import { OmitReferences, Producto as ProductoType } from '@/data/types'
import { cn } from '@/lib/utils'
import Image from 'next/image'
import Link from 'next/link'
import Heading from './Heading'
import { Waypoint } from 'react-waypoint'
import { useState } from 'react'

interface ProductoProps extends OmitReferences<ProductoType> {
  concavo?: boolean
  className?: string
}

export const Producto = ({ concavo = false, className, ...producto }: ProductoProps) => {
  const [animate, setAnimate] = useState(false)

  return (
    <>
      <Link
        href={`/productos/${producto.slug}`}
        className={cn(className, 'cursor-pointer text-dark-gray group')}
        aria-label={producto.name}
      >
        <Waypoint
          onEnter={() => {
            setAnimate(true)
          }}
          bottomOffset='20%'
        />
        <div className='pt-[100%] relative overflow-hidden '>
          <Image
            fill
            src={`/productos/${producto.slug}.jpg`}
            alt={producto.name}
            className={cn('object-cover group-hover:scale-105', animate ? 'opacity-100' : 'opacity-0')}
          />
        </div>
        <div className='mb-4 group-hover:text-light-gray transition-colors duration-500'>
          <Heading
            className='text-sm xs:text-base sm:text-lg mt-4 mb-0 sm:mt-6 sm:mb-0 md:mt-8 md:mb-0'
            animated={false}
          >
            {producto.name}
          </Heading>
          {concavo && (
            <p className='text-xs text-center mt-0 leading-none' aria-label='dimensions'>
              AAA x LLL x PPP
            </p>
          )}
        </div>
      </Link>
    </>
  )
}
