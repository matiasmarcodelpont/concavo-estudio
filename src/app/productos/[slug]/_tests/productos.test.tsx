import '@testing-library/jest-dom'
import { render, screen, within } from '@testing-library/react'

import Producto from '../page'

jest.mock('@/data/data.ts', () => ({
  ambientes: [
    {
      slug: 'living-comedor',
      name: 'Living/Comedor',
      productos: [
        {
          slug: 'lampara-hierro',
        },
        {
          slug: 'piso-madera',
        },
      ],
      colaboradores: [],
    },
    {
      slug: 'escritorio',
      name: 'Escritorio',
      productos: [
        {
          slug: 'lampara-hierro',
        },
        {
          slug: 'piso-madera',
        },
      ],
      colaboradores: [],
    },
    {
      slug: 'cuarto-principal',
      name: 'Cuarto principal',
      productos: [
        {
          slug: 'lampara-hierro',
        },
        {
          slug: 'sofa-carlos',
        },
      ],
      colaboradores: [],
    },
  ],
  productos: [
    {
      slug: 'piso-madera',
      name: 'Piso simil Madera',
      colaborador: {
        slug: 'elegance',
      },
    },
    {
      slug: 'lampara-hierro',
      name: 'Lámpara de diseño con detalles de Hierro',
      colaborador: {
        slug: 'luz-viva',
      },
    },
    {
      slug: 'sofa-carlos',
      name: 'Sofá Carlos V',
      colaborador: {
        slug: 'muebles-arte',
      },
    },
  ],
  colaboradores: [],
}))

describe('Producto', () => {
  it('renders producto page unchanged', () => {
    const { container } = render(<Producto params={{ productoSlug: 'lampara-hierro' }} />)
    expect(container).toMatchSnapshot()
  })

  it('renders Producto name', () => {
    render(<Producto params={{ productoSlug: 'lampara-hierro' }} />)

    const heading = screen.getByRole('heading', { level: 1, name: 'Lámpara de diseño con detalles de Hierro' })
    expect(heading).toBeInTheDocument()
  })

  it('renders related productos', () => {
    render(<Producto params={{ productoSlug: 'lampara-hierro' }} />)

    const productosList = screen.getByRole('list', { name: 'Lista de productos relacionados' })
    expect(productosList).toBeInTheDocument()

    const productos = within(productosList).getAllByRole('listitem')
    expect(productos).toHaveLength(2)

    expect(productos[0]).toBeInTheDocument()
    const link0 = within(productos[0]).getByRole<HTMLAnchorElement>('link', { name: 'Piso simil Madera' })
    expect(link0).toBeInTheDocument()
    expect(link0.href).toMatch(/^https?:\/\/[^/]+\/productos\/piso-madera$/)

    expect(productos[1]).toBeInTheDocument()
    const link1 = within(productos[1]).getByRole<HTMLAnchorElement>('link', { name: 'Sofá Carlos V' })
    expect(link1).toBeInTheDocument()
    expect(link1.href).toMatch(/^https?:\/\/[^/]+\/productos\/sofa-carlos$/)
  })

  it('shows the 404 page if producto is not found', () => {
    expect(() => {
      render(<Producto params={{ productoSlug: 'productoo-inexistente' }} />)
    }).toThrow('NEXT_NOT_FOUND')
  })
})
