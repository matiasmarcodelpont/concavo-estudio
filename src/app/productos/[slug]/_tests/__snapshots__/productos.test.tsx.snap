// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Producto renders producto page unchanged 1`] = `
<div>
  <main>
    <span
      style="font-size: 0px;"
    />
    <h1
      class="uppercase text-3xl mt-12 mb-6 sm:mt-16 sm:mb-9 md:mt-20 md:mb-12 italic opacity-0 text-center md:hidden"
    >
      Lámpara de diseño con detalles de Hierro
    </h1>
    <div
      class="mb-16 flex flex-wrap"
    >
      <div
        aria-roledescription="carousel"
        class="relative mx-auto h-[61.8vw] max-h-[494px] w-full md:w-1/2"
        role="region"
      >
        <div
          class="overflow-hidden h-full"
        >
          <div
            class="flex h-full"
            style="transform: translate3d(0px,0px,0px);"
          >
            <div
              aria-roledescription="slide"
              class="min-w-0 shrink-0 grow-0 basis-full relative"
              role="group"
            >
              <img
                alt="Lámpara de diseño con detalles de Hierro"
                class="object-cover"
                data-nimg="fill"
                decoding="async"
                loading="lazy"
                sizes="100vw"
                src="/_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=3840&q=75"
                srcset="/_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=640&q=75 640w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=750&q=75 750w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=828&q=75 828w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=3840&q=75 3840w"
                style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
              />
            </div>
            <div
              aria-roledescription="slide"
              class="min-w-0 shrink-0 grow-0 basis-full relative"
              role="group"
            >
              <img
                alt="Lámpara de diseño con detalles de Hierro"
                class="object-cover"
                data-nimg="fill"
                decoding="async"
                loading="lazy"
                sizes="100vw"
                src="/_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=3840&q=75"
                srcset="/_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=640&q=75 640w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=750&q=75 750w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=828&q=75 828w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=3840&q=75 3840w"
                style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
              />
            </div>
            <div
              aria-roledescription="slide"
              class="min-w-0 shrink-0 grow-0 basis-full relative"
              role="group"
            >
              <img
                alt="Lámpara de diseño con detalles de Hierro"
                class="object-cover"
                data-nimg="fill"
                decoding="async"
                loading="lazy"
                sizes="100vw"
                src="/_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=3840&q=75"
                srcset="/_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=640&q=75 640w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=750&q=75 750w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=828&q=75 828w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fproductos%2Flampara-hierro.jpg&w=3840&q=75 3840w"
                style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
              />
            </div>
          </div>
        </div>
        <button
          class="absolute transition-opacity left-0 top-1/2 -translate-y-1/2"
        >
          <svg
            class="lucide lucide-chevron-left "
            fill="none"
            height="50px"
            stroke="white"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="50px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m15 18-6-6 6-6"
            />
          </svg>
          <span
            class="sr-only"
          >
            Previous slide
          </span>
        </button>
        <button
          class="absolute transition-opacity right-0 top-1/2 -translate-y-1/2"
        >
          <svg
            class="lucide lucide-chevron-right "
            fill="none"
            height="50px"
            stroke="white"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="50px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m9 18 6-6-6-6"
            />
          </svg>
          <span
            class="sr-only"
          >
            Next slide
          </span>
        </button>
      </div>
      <section
        class="px-4 sm:px-6 md:px-8 mt-10 md:w-1/2"
      >
        <span
          style="font-size: 0px;"
        />
        <h1
          class="uppercase text-3xl mb-6 sm:mb-9 md:mb-12 italic opacity-0 text-center mt-0 sm:mt-0 md:mt-0 hidden md:block"
        >
          Lámpara de diseño con detalles de Hierro
        </h1>
        <p />
      </section>
    </div>
    <section
      class="mx-6 sm:mx-10 md:mx-12 mb-12 text-center"
    >
      <span
        style="font-size: 0px;"
      />
      <h1
        class="uppercase mt-12 mb-6 sm:mt-16 sm:mb-9 md:mt-20 md:mb-12 italic opacity-0 text-xl sm:text-2xl md:text-3xl"
      >
        Más productos
      </h1>
      <ul
        aria-label="Productos relacionados"
        class="grid grid-cols-auto-fit-100 xs:grid-cols-auto-fit-150 sm:grid-cols-auto-fill-200 md:grid-cols-auto-fill-300 gap-4"
      >
        <li>
          <a
            aria-label="Piso simil Madera"
            class="cursor-pointer text-dark-gray group"
            href="/productos/piso-madera"
          >
            <span
              style="font-size: 0px;"
            />
            <div
              class="pt-[100%] relative overflow-hidden "
            >
              <img
                alt="Piso simil Madera"
                class="object-cover group-hover:scale-105 opacity-0"
                data-nimg="fill"
                decoding="async"
                loading="lazy"
                sizes="100vw"
                src="/_next/image?url=%2Fproductos%2Fpiso-madera.jpg&w=3840&q=75"
                srcset="/_next/image?url=%2Fproductos%2Fpiso-madera.jpg&w=640&q=75 640w, /_next/image?url=%2Fproductos%2Fpiso-madera.jpg&w=750&q=75 750w, /_next/image?url=%2Fproductos%2Fpiso-madera.jpg&w=828&q=75 828w, /_next/image?url=%2Fproductos%2Fpiso-madera.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fproductos%2Fpiso-madera.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fproductos%2Fpiso-madera.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fproductos%2Fpiso-madera.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fproductos%2Fpiso-madera.jpg&w=3840&q=75 3840w"
                style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
              />
            </div>
            <div
              class="mb-4 group-hover:text-light-gray transition-colors duration-500"
            >
              <span
                style="font-size: 0px;"
              />
              <h1
                class="uppercase italic opacity-100 text-sm xs:text-base sm:text-lg mt-4 mb-0 sm:mt-6 sm:mb-0 md:mt-8 md:mb-0"
              >
                Piso simil Madera
              </h1>
            </div>
          </a>
        </li>
        <li>
          <a
            aria-label="Sofá Carlos V"
            class="cursor-pointer text-dark-gray group"
            href="/productos/sofa-carlos"
          >
            <span
              style="font-size: 0px;"
            />
            <div
              class="pt-[100%] relative overflow-hidden "
            >
              <img
                alt="Sofá Carlos V"
                class="object-cover group-hover:scale-105 opacity-0"
                data-nimg="fill"
                decoding="async"
                loading="lazy"
                sizes="100vw"
                src="/_next/image?url=%2Fproductos%2Fsofa-carlos.jpg&w=3840&q=75"
                srcset="/_next/image?url=%2Fproductos%2Fsofa-carlos.jpg&w=640&q=75 640w, /_next/image?url=%2Fproductos%2Fsofa-carlos.jpg&w=750&q=75 750w, /_next/image?url=%2Fproductos%2Fsofa-carlos.jpg&w=828&q=75 828w, /_next/image?url=%2Fproductos%2Fsofa-carlos.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fproductos%2Fsofa-carlos.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fproductos%2Fsofa-carlos.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fproductos%2Fsofa-carlos.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fproductos%2Fsofa-carlos.jpg&w=3840&q=75 3840w"
                style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
              />
            </div>
            <div
              class="mb-4 group-hover:text-light-gray transition-colors duration-500"
            >
              <span
                style="font-size: 0px;"
              />
              <h1
                class="uppercase italic opacity-100 text-sm xs:text-base sm:text-lg mt-4 mb-0 sm:mt-6 sm:mb-0 md:mt-8 md:mb-0"
              >
                Sofá Carlos V
              </h1>
            </div>
          </a>
        </li>
      </ul>
    </section>
  </main>
</div>
`;
